ethportal-api/src/types/ping_extensions/extensions/type_0.rs

```diff
@@ -0,0 +1,213 @@
+            self.cpu_architecture,
+            self.programming_language_version
+        )
+    }
+}
+
+impl FromStr for ClientInfo {
+    type Err = anyhow::Error;
+
+    fn from_str(string: &str) -> Result<Self, anyhow::Error> {
+        let parts: Vec<&str> = string.split('/').collect();
+
+        if parts.len() != 4 {
+            bail!("Invalid client info string: should have 4 /'s {}", string);
+        }
+
+        let client_name = parts[0];
+
+        let Some((client_version, short_commit)) = parts[1].split('-').collect_tuple() else {
+            bail!(
+                "Invalid client info string: should look like 0.1.1-2b00d730 got {}",
+                parts[1]
+            );
+        };
+
+        let Some((operating_system, cpu_architecture)) = parts[2].split('-').collect_tuple() else {
+            bail!(
+                "Invalid client info string: should look like linux-x86_64 got {}",
+                parts[2]
+            );
+        };
+
+        Ok(Self {
+            client_name: client_name.to_string(),
+            client_version: client_version.to_string(),
+            short_commit: short_commit.to_string(),
+            operating_system: operating_system.to_string(),
+            cpu_architecture: cpu_architecture.to_string(),
+            programming_language_version: parts[3].to_string(),
+        })
+    }
+}
+
+impl Encode for ClientInfo {
+    fn is_ssz_fixed_len() -> bool {
+        false
+    }
+
+    fn ssz_append(&self, buf: &mut Vec<u8>) {
+        let bytes: Vec<u8> = self.string().as_bytes().to_vec();
+        let byte_list: VariableList<u8, U200> = VariableList::from(bytes);
+        buf.extend_from_slice(&byte_list);
+    }
+
+    fn ssz_bytes_len(&self) -> usize {
+        self.as_ssz_bytes().len()
+    }
+}
+
+impl Decode for ClientInfo {
+    fn from_ssz_bytes(bytes: &[u8]) -> Result<Self, ssz::DecodeError> {
+        let byte_list = VariableList::<u8, U200>::from_ssz_bytes(bytes)?;
+        let string = String::from_utf8(byte_list.to_vec()).map_err(|_| {
+            ssz::DecodeError::BytesInvalid(format!("Invalid utf8 string: {byte_list:?}"))
+        })?;
+        Self::from_str(&string).map_err(|err| {
+            ssz::DecodeError::BytesInvalid(format!("Failed to parse client info: {err:?}"))
+        })
+    }
+
+    fn is_ssz_fixed_len() -> bool {
+        false
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_client_info_round_trip() {
+        let client_info = ClientInfo::trin_client_info();
+        let bytes = client_info.as_ssz_bytes();
+        let decoded = ClientInfo::from_ssz_bytes(&bytes).unwrap();
+        assert_eq!(client_info, decoded);
+    }
+
+    #[test]
+    fn test_client_info_from_str() {
+        let client_info = ClientInfo::trin_client_info();
+        let string = client_info.string();
+        let decoded = ClientInfo::from_str(&string).unwrap();
+        assert_eq!(client_info, decoded);
+    }
+
+    #[test]
```

Comment:
Seems like you could use rstest for these fail cases, and just add inline comments to the individual cases explaining why they should fail